src/batch/src/executor2/generic_exchange.rs

```diff
@@ -128,60 +136,77 @@ impl<CS: 'static + CreateSource> Executor2 for GenericExchangeExecutor2<CS> {
 
 impl<CS: 'static + CreateSource> GenericExchangeExecutor2<CS> {
     #[try_stream(boxed, ok = DataChunk, error = RwError)]
-    async fn do_execute(mut self: Box<Self>) {
-        loop {
-            if self.source_idx >= self.sources.len() {
-                break;
-            }
-            if self.current_source.is_none() {
-                let proto_source = &self.sources[self.source_idx];
-                let source =
-                    CS::create_source(self.env.clone(), proto_source, self.task_id.clone()).await?;
-                self.current_source = Some(source);
+    async fn do_execute(self: Box<Self>) {
+        let mut sources: Vec<Box<dyn ExchangeSource>> = vec![];
+
+        for prost_source in &self.sources {
+            let source =
+                CS::create_source(self.env.clone(), prost_source, self.task_id.clone()).await?;
+            sources.push(source);
+        }
+
+        let mut handles = sources
+            .into_iter()
+            .map(data_chunk_stream)
+            .map(|stream| stream.into_future())
+            .collect_vec();
+
+        while !handles.is_empty() {
+            let ((result, from), _id, remainings) = select_all(handles).await;
+            handles = remainings;
+
+            if let Some(data_chunk) = result {
+                let data_chunk = data_chunk?;
+                handles.push(from.into_future());
+                yield data_chunk
             }
-            let mut source = self.current_source.take().unwrap();
-            match source.take_data().await? {
-                None => {
-                    self.current_source = None;
-                    self.source_idx += 1;
-                }
-                Some(res) => {
-                    if res.cardinality() == 0 {
-                        debug!("Exchange source {:?} output empty chunk.", source);
-                        assert_ne!(res.cardinality(), 0);
-                    }
-                    self.current_source = Some(source);
-                    yield res;
-                }
+        }
+    }
+}
+#[try_stream(boxed, ok = DataChunk, error = RwError)]
+async fn data_chunk_stream(mut source: Box<dyn ExchangeSource>) {
+    loop {
+        if let Some(res) = source.take_data().await? {
+            if res.cardinality() == 0 {
+                debug!("Exchange source {:?} output empty chunk.", source);
+                assert_ne!(res.cardinality(), 0);
```

Comment:
```suggestion
```

Is this assertion unnecessary?

Reply:
The original logic was written that way so I kept it that way.

Reply:
I will remove that